drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -679,6 +702,62 @@ def has_submission_deadline_passed(self):
         else:
             return False
 
+    @property
+    def closed(self):
+        """
+        Is the student still allowed to submit answers?
+        """
+        if not self.attempts_remain:
+            return True
+        if self.has_submission_deadline_passed:
+            return True
+
+        return False
+
+    @property
+    def is_attempted(self):
+        """
+        Has the problem been attempted?
+        """
+        return self.attempts > 0
+
+    @property
+    def is_finished(self):
+        """
+        Returns True if answer is closed or answer is correct.
+        """
+        return self.closed or self.is_correct
+
+    @property
+    def is_answer_available(self):
+        """
+        Is student allowed to see an answer?
+        """
+        _look_up = {
+            SHOWANSWER.NEVER: lambda: False,
+            SHOWANSWER.ATTEMPTED: lambda: self.is_attempted or self.has_submission_deadline_passed,
+            SHOWANSWER.ANSWERED: lambda: self.is_correct,
+            SHOWANSWER.CLOSED: lambda: self.closed,
+            SHOWANSWER.FINISHED: lambda: self.is_finished,
+            SHOWANSWER.CORRECT_OR_PAST_DUE: lambda: self.is_correct or self.has_submission_deadline_passed,
+            SHOWANSWER.PAST_DUE: lambda: self.has_submission_deadline_passed,
+            SHOWANSWER.ALWAYS: lambda: True,
+            SHOWANSWER.AFTER_ALL_ATTEMPTS: lambda: not self.attempts_remain,
+            SHOWANSWER.AFTER_ALL_ATTEMPTS_OR_CORRECT: lambda: not self.attempts_remain or self.is_correct,
+            SHOWANSWER.ATTEMPTED_NO_PAST_DUE: lambda: self.is_attempted,
+        }
+
+        if self.mode != Constants.ASSESSMENT_MODE:
+            return False
+        current_user = self.runtime.service(self, 'user').get_current_user()
+        user_is_staff = current_user.opt_attrs.get(Constants.ATTR_KEY_USER_IS_STAFF)
```

Comment:
We should check if the service exists. There are some rare edge cases where it can be absent.
For example, it can happen when a user tries to access a Course Outline page of a course that doesn't have its block structures generated yet.